File: src/graph/bfs/km105.js
// @ts-check
import { createInterface } from "readline";

let /** @type {number} */ nodeNum = 0;
let /** @type {number} */ edgeNum = 0;

const /** @type {Map<number, number[]>} */ graph = new Map();

let /** @type {number} */ lineno = 0;

const rl = createInterface({
  input: process.stdin,
  output: process.stdout,
});

rl.on("line", (line) => {
  lineno++;
  if (lineno === 1) {
    [nodeNum, edgeNum] = line
      .trim()
      .split(" ")
      .map((item) => Number.parseInt(item));
    return;
  }

  const [start, end] = line
    .trim()
    .split(" ")
    .map((item) => Number.parseInt(item));

  if (graph.has(start)) {
    graph.get(start)?.push(end);
  } else {
    graph.set(start, [end]);
  }

  if (lineno === edgeNum + 1) {
    rico();
    rl.close();
  }
});

const /** @type {boolean[]} */ visit = new Array(nodeNum + 1).fill(false);

/**
 *
 * @param {number} key
 */
function dfs(key) {
  visit[key] = true;

  if (graph.has(key)) {
    for (const sibling of graph.get(key) ?? []) {
      if (!visit[sibling]) {
        dfs(sibling);
      }
    }
  }
}

function rico() {
  dfs(1);
  for (const item of visit.slice(1)) {
    if (!item) {
      process.stdout.write("-1\n");
      process.exit(0);
    }
  }

  process.stdout.write("1\n");
}

File: src/graph/dfs/km105.py
from typing import Dict, List


nodeNum, edgeNum = map(int, input().strip().split(" "))

graph: Dict[int, List[int]] = dict()

for _ in range(edgeNum):
    start, end = map(int, input().strip().split(" "))
    if start in graph:
        graph[start].append(end)
    else:
        graph[start] = [end]

visit: List[bool] = [False for _ in range(nodeNum + 1)]


def dfs(key: int) -> None:
    visit[key] = True

    if key in graph:
        for sibling in graph[key]:
            if not visit[sibling]:
                dfs(sibling)


dfs(1)
for item in visit[1:]:
    if not item:
        print(-1)
        exit(0)

print(1)
